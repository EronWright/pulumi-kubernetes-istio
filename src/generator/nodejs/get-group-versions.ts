import { GroupVersions } from './interfaces';

export function getGroupVersions(paths: string[]): GroupVersions[] {
  return paths
    .map(path => {
      const paths = path.split('/');
      return paths.slice(0, paths.length - 1);
    })
    .filter(urn => urn.length < 3)
    .map(urn => ({
      group: urn[0],
      versions: [urn[1]],
    }))
    .reduce((items, item) => {
      const group = items.find(({ group }) => group === item.group);

      if (group) {
        group.versions = [...new Set([...group.versions, ...item.versions])];
      }

      return group ? items : [...items, item];
    }, []);
}
