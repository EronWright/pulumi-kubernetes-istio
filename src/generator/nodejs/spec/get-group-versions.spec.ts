import { getGroupVersions } from '../get-group-versions';

describe('getGroupVersions', () => {
  it('should combine all paths into group -> versions with no duplicates', () => {
    const paths = [
      'authentication/v1alpha1/istio.authentication.v1alpha1.json',
      'networking/v1alpha3/destination_rule.json',
      'networking/v1alpha3/envoy_filter.json',
      'policy/v1beta1/istio.policy.v1beta1.json',
      'rbac/v1alpha1/istio.rbac.v1alpha1.json',
      'rbac/v1alpha2/istio.rbac.v1alpha1.json',
      'mixer/adapter/model/v1beta1/istio.mixer.adapter.model.v1beta1.json',
      'mixer/v1/config/client/istio.mixer.v1.config.client.json',
    ];

    const groupVersions = getGroupVersions(paths);

    expect(groupVersions).toStrictEqual([
      {
        group: 'authentication',
        versions: ['v1alpha1'],
      },
      {
        group: 'networking',
        versions: ['v1alpha3'],
      },
      {
        group: 'policy',
        versions: ['v1beta1'],
      },
      {
        group: 'rbac',
        versions: ['v1alpha1', 'v1alpha2'],
      },
    ]);

    expect(groupVersions).toMatchSnapshot();
  });
});
