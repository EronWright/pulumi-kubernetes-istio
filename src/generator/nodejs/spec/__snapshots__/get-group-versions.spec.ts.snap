// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`getGroupVersions should combine all paths into a group -> versions with no duplicates 1`] = `
Array [
  Object {
    "group": "authentication",
    "versions": Array [
      "v1alpha1",
    ],
  },
  Object {
    "group": "networking",
    "versions": Array [
      "v1alpha3",
    ],
  },
  Object {
    "group": "policy",
    "versions": Array [
      "v1beta1",
    ],
  },
  Object {
    "group": "rbac",
    "versions": Array [
      "v1alpha1",
      "v1alpha2",
    ],
  },
]
`;
